Exercice_1_-_Le_champ_de_Noel/champ.py

```diff
@@ -32,19 +32,23 @@
 BORDER_CHAR = "#"
 EMPTY_CHAR = " "
 TREE_DESIGN = ["^", "^ ^", "(o  )", "(o  o )", "U"]
+# empiric factor to reduce tree spawn occurrence
+TREE_OCCURRENCE_FACTOR = 0.25
 
 
 def main():
-    # initialization
     # create field with borders
     field = [
         [BORDER_CHAR if (x in X_LIMITS or y in Y_LIMITS) else EMPTY_CHAR for x in range(WIDTH)]
         for y in range(HEIGHT)
     ]
 
-    # empiric max number of trees for the sake of visibility
-    # WIDTH and HEIGHT sizes contribute to the number of trees
-    max_tree_for_field = math.ceil((WIDTH * HEIGHT) ** (1 / 2) / 4)
+    # Empiric max number of trees for the sake of visibility.
+    # My goal was to have a dynamic approach for having "enough" trees to demonstrate perspective
+    # but also to have "enough" probability to see a clean field with few trees.
+    # Thus, both WIDTH and HEIGHT sizes contribute to the maximum number of trees, multiplied by
+    # a coefficient TREE_OCCURRENCE_FACTOR to reduce the spawn occurrence of trees.
+    max_tree_for_field = int((WIDTH + HEIGHT) * TREE_OCCURRENCE_FACTOR)
```

Comment:
Maintenant qu'elle est bien expliquée... je peux voir que cette formule est fausse :smile: 

Tu veux éviter d'avoir trop d'arbre. Mais tout ce que tu sais c'est que `max_tree_for_field` doit te donner ton nombre max d'arbre. Mais ton analyse est fausse... certe ça augmente avec la taille du champ aussi bien en x qu'en y mais dans ce cas là il faudrait un multiplication des deux (si si je te jure c'est une question d'analyse aux dimensions...). En gros c'est proportionnel à la surface.

Mais surtout ça dépend de la surface de ton arbre ! Plus l'arbre prend de place moins tu peux en mettre. Il nous faut donc les dimensions de l'arbre :

```python
TREE_HEIGHT = len(TREE_DESIGN)
TREE_WIDTH = max([len(line] for line in TREE_DESIGN)
```

Maintenant ta formule devient (en majuscule au passage vu que c'est une constante !!!!) :

```python
MAX_TREE_FOR_FIELD = int((WIDTH + HEIGHT) /  (TREE_WIDTH * TREE_HEIGHT))
```

On voit que si l'arbre et le champ font la même taille on ne peut mettre qu'un seul arbre dans le champ en gros. Et niveau dimension ça colle bien : [m]²/[m]² --> nombre. Nickel.

Maintenant comment ajouter un facteur là-dessus pour qu'on puisse contrôler tout ça ? En fait ce qu'on vient de calculer c'est le nombre d'arbre maximum avec une densité de 100% (toute la surface est occupée par des arbre en moyenne). Il suffit de multiplier ça par un nombre entre 0.0 et 1.0 (càd un pourcentage) qui représente le taux d'occupation du sol par les arbre (`TREE_OCCUPATION_RATE`) qui est sans dimension : 

- 0.5 --> la moitié de la surface est couverte en moyenne
- 0.1 --> 10% de la surface est couverte (c'est une prairie avec qq arbres)
- 0.9 --> 90% de la surface est couverte d'arbre (c'est une forêt assez dense)
- 1.5 --> 150% de la surface est couverte donc il y a pus d'arbre que le max, on a forcément des recouvrement dans 50% des cas (en moyenne) c'est une forêt très dense

(:point_up: voilà le genre de commentaire hyper utile à mettre au dessus d'une constante de ce genre : explication de sa signification + exemples clairs avec interprétation)

Ta formule devrait donc être :

```python
MAX_TREE_FOR_FIELD = floor(TREE_OCCUPATION_RATE * (WIDTH + HEIGHT) /  (TREE_WIDTH * TREE_HEIGHT))
```

Tu notera que j'ai utilisé `floor` du module `math` plutôt que `int` car on calcul un nombre maximum : si on a la place pour 12.7 arbres, en nombre entier on a la place pour 12 arbre. Ça doit être explicite.